src/sinks/util/http.rs

```diff
@@ -598,6 +599,144 @@ pub fn validate_headers(
+        self.finalizers.take_finalizers()
+    }
+}
+
+impl MetaDescriptive for HttpRequest {
+    fn get_metadata(&self) -> &RequestMetadata {
+        &self.request_metadata
+    }
+
+    fn metadata_mut(&mut self) -> &mut RequestMetadata {
+        &mut self.request_metadata
+    }
+}
+
+impl ByteSizeOf for HttpRequest {
+    fn allocated_bytes(&self) -> usize {
+        self.payload.allocated_bytes() + self.finalizers.allocated_bytes()
+    }
+}
+
+/// Response type for use in the `Service` implementation of HTTP stream sinks.
+pub struct HttpResponse {
+    pub http_response: Response<Bytes>,
+    events_byte_size: GroupedCountByteSize,
+    raw_byte_size: usize,
+}
+
+impl DriverResponse for HttpResponse {
+    fn event_status(&self) -> EventStatus {
+        if self.http_response.status().is_success() {
+            EventStatus::Delivered
+        } else {
+            EventStatus::Rejected
+        }
+    }
+
+    fn events_sent(&self) -> &GroupedCountByteSize {
+        &self.events_byte_size
+    }
+
+    fn bytes_sent(&self) -> Option<usize> {
+        Some(self.raw_byte_size)
+    }
+}
+
+/// HTTP request builder for batched HTTP stream sinks using the generic `HttpService`
+pub trait HttpServiceRequestBuilder {
+    fn build(&self, body: Bytes) -> Request<Bytes>;
+}
+
+/// Generic 'Service' implementation for batched HTTP stream sinks.
+#[derive(Clone)]
+pub struct HttpService<B> {
+    batch_service:
+        HttpBatchService<BoxFuture<'static, Result<Request<Bytes>, crate::Error>>, HttpRequest>,
+    _phantom: PhantomData<B>,
+}
+
+impl<B> HttpService<B>
+where
+    B: HttpServiceRequestBuilder + std::marker::Sync + std::marker::Send + 'static,
+{
+    pub fn new(http_client: HttpClient<Body>, http_request_builder: B) -> Self {
+        let http_request_builder = Arc::new(http_request_builder);
+
+        let batch_service = HttpBatchService::new(http_client, move |req| {
+            let req: HttpRequest = req;
+
+            let request_builder = Arc::clone(&http_request_builder);
+
+            let fut: BoxFuture<'static, Result<http::Request<Bytes>, crate::Error>> =
+                Box::pin(async move { Ok(request_builder.build(req.payload)) });
+
+            fut
+        });
+        Self {
+            batch_service,
+            _phantom: PhantomData,
+        }
+    }
+}
+
+impl<B> Service<HttpRequest> for HttpService<B>
+where
+    B: HttpServiceRequestBuilder + std::marker::Sync + std::marker::Send + 'static,
+{
+    type Response = HttpResponse;
+    type Error = crate::Error;
+    type Future = BoxFuture<'static, Result<Self::Response, Self::Error>>;
+
+    fn poll_ready(&mut self, _cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        Poll::Ready(Ok(()))
+    }
+
+    fn call(&mut self, mut request: HttpRequest) -> Self::Future {
+        let mut http_service = self.batch_service.clone();
+
+        let raw_byte_size = request.payload.len();
+        let metadata = std::mem::take(request.metadata_mut());
```

Comment:
<!-- Thoughts represent an idea that popped up from reviewing. These comments are non-blocking by nature, but they are extremely valuable and can lead to more focused initiatives and mentoring opportunities.  -->
**💭 thought:** ‏

This looks fine,  but could hide future bugs. After this point `request` will no longer have it's metadata which is something that won't be obvious to `http_service.call`. Perhaps a comment here to warn about this may help any future coders from tripping over this.